listenbrainz/db/year_in_music.py

```diff
@@ -227,6 +228,123 @@ def insert_playlists(troi_patch_slug, import_file):
         current_app.logger.error("Error while inserting playlist/%s:" % troi_patch_slug, exc_info=True)
 
 
+def caa_id_to_archive_url(release_mbid, caa_id):
+    return f"https://archive.org/download/mbid-{release_mbid}/mbid-{release_mbid}-{caa_id}_thumb500.jpg"
+
+
+def get_coverart_for_top_releases(top_releases):
+    """Use the CAA database connection to find coverart for each release
+    1. For all releases, find coverart
+    2. For releases where there is no coverart, get their releasegroup
+    3. Find release mbids for the above release groups
+    4. For these release mbids, find coverart
+    """
+
+    release_mbids = [rel["release_mbid"] for rel in top_releases if "release_mbid" in rel]
+
+    release_coverart = {}
+    release_mbid_to_row = {}
+    release_id_to_mbid = {}
+    release_mbid_to_release_group_id = {}
+    with musicbrainz_db.engine.connect() as connection:
+        query = """SELECT release_gid_redirect.gid::text
+                        , new_id
+                        , release.release_group
+                     FROM release_gid_redirect
+                     JOIN release
+                       ON release.id = new_id
+                    WHERE release_gid_redirect.gid IN :release_mbids"""
+        res = connection.execute(sqlalchemy.text(query), {"release_mbids": tuple(release_mbids)})
+        for row in res.fetchall():
+            release_mbid_to_row[row["gid"]] = row["new_id"]
+            release_id_to_mbid[row["new_id"]] = row["gid"]
+            release_mbid_to_release_group_id[row["gid"]] = row["release_group"]
+        query = """SELECT gid::text
+                        , id
+                        , release_group
+                     FROM release
+                    WHERE gid IN :release_mbids"""
+        res = connection.execute(sqlalchemy.text(query), {"release_mbids": tuple(release_mbids)})
+        for row in res.fetchall():
+            release_mbid_to_row[row["gid"]] = row["id"]
+            release_id_to_mbid[row["id"]] = row["gid"]
+            release_mbid_to_release_group_id[row["gid"]] = row["release_group"]
+
+        # We need to know what the release mbid was that we passed in which gave us this rgid
+        release_group_id_to_release_mbid = {v: k for k, v in release_mbid_to_release_group_id.items()}
+
+        # Front coverart
+        query = """SELECT id
+                        , release 
+                     FROM cover_art_archive.index_listing
+                    WHERE release in :release_ids
+                      AND is_front = 't';
+        """
+        res = connection.execute(sqlalchemy.text(query), {"release_ids": tuple(release_id_to_mbid.keys())})
+        for row in res.fetchall():
+            release_id = row["release"]
+            caa_id = row["id"]
+            release_mbid = release_id_to_mbid[release_id]
```

Comment:
there are a handful of places where I'm getting data out of a dict that I'm pretty sure is always going to be there, but I don't know if it might fail in some unexpected situations